### PR TITLE
deja/robot_name

### PR DESCRIPTION
**Pull request by DejaAugustine**
Incorporated latest feedback and split into separate feature branch.

### CODE REVIEW

File: robot_name/registry.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+require "set"
```
Comment (paulghaddad) [inline robot_name/registry.rb]: I like your use of `Set`. Most people don't realize it exists!

File: robot_name/registry.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+require_relative "errors.rb"
+
+class RobotRegistry
+  def initialize(args = {})
+    @registry = args[:registry] || Set.new
+  end
+
+  def add(name)
+    raise NameCollisionError,
```
Comment (paulghaddad) [inline robot_name/registry.rb]: I like having your error conditional right here.

File: robot_name/registry.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      "#{name} is already in the registry!" if @registry.include?(name)
```
Comment (paulghaddad) [inline robot_name/registry.rb]: Nice job providing a detailed error message as well.

File: robot_name/registry.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+    @registry << name
+  end
+
+  alias_method :<<, :add
```
Comment (paulghaddad) [inline robot_name/registry.rb]: :+1:

File: robot_name/registry.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+require "set"
+require_relative "errors.rb"
```
Comment (paulghaddad) [inline robot_name/registry.rb]: Omitting the `.rb` should work as well:

`require_relative "errors"`

File: robot_name/robot_name.rb [hunk 6]
```diff
@@ -1,32 +1,93 @@
-class NameCollisionError < RuntimeError; end
+require_relative "errors.rb"
+require_relative "registry.rb"
 
 class Robot
-  attr_accessor :name
-
-  @@registry
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: Good job getting rid of the class variable.

File: robot_name/registry.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+class RobotRegistry
```
Comment (paulghaddad) [inline robot_name/registry.rb]: I really like this class. Short, simple, easy to read with a single responsibility.

File: robot_name/robot_name.rb [hunk 8]
```diff
@@ -1,32 +1,93 @@
...
+  attr_reader :name, :registry
+  attr_reader :name_generator, :name_validator
 
   def initialize(args = {})
-    @@registry ||= []
-    @name_generator = args[:name_generator]
+    @name_generator = args[:name_generator] || method(:generate_name)
+    @name_validator = args[:name_validator] || method(:valid_name?)
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: I really like the idioms in Lines 9 and 10.

File: robot_name/robot_name.rb [hunk 9]
```diff
@@ -1,32 +1,93 @@
...
+
+  def validate
+    raise NameValidationError,
+      "#{name} is not a valid name!" unless name_validator.call(name)
+  end
+
+  def generate_name
+    new_name = ""
+
+    2.times { new_name.concat(('A'..'Z').to_a.sample) }
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: The only thing about lines 30 and 31 is to consider extracting their functionality into private methods.

File: robot_name/robot_name.rb [hunk 10]
```diff
@@ -1,32 +1,93 @@
...
+  def register
+    registry << name
+  end
+
+  def validate
+    raise NameValidationError,
+      "#{name} is not a valid name!" unless name_validator.call(name)
+  end
+
+  def generate_name
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: This is a nice method, but see if using `inject` to build up your string would work better or not.

File: robot_name/robot_name.rb [hunk 11]
```diff
@@ -1,32 +1,93 @@
...
+puts ""
+puts "***********************************"
+puts "* Robot Name Generator & Registry *"
+puts "***********************************"
+puts ""
+
+registry = RobotRegistry.new
+
+# Demonstrate Generator
+begin
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: I like this! It's the first time I've seen a `rescue` used here.

File: robot_name/robot_name.rb [hunk 12]
```diff
@@ -1,32 +1,93 @@
...
+    puts "Robot '#{robot.name}' Registered! (default generator and validator)"
+  end
+rescue RobotError => e
+  puts "Error Generating Robot: #{e.message}"
+end
 
-    if @name_generator
-      @name = @name_generator.call
-    else
-      generate_char = -> { ('A'..'Z').to_a.sample }
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: Good job getting rid of the lambdas. They are not needed.

File: robot_name/robot_name.rb [hunk 13]
```diff
@@ -1,32 +1,93 @@
...
+
+    validate
+    register
+  end
+
+  def register
+    registry << name
+  end
+
+  def validate
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: This method can be private.

File: robot_name/robot_name.rb [hunk 14]
```diff
@@ -1,32 +1,93 @@
...
+    @name_generator = args[:name_generator] || method(:generate_name)
+    @name_validator = args[:name_validator] || method(:valid_name?)
+    @registry = args[:registry]
+    @name = name_generator.call
+
+    validate
+    register
+  end
+
+  def register
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: This method can also be made private.

File: robot_name/registry.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  def initialize(args = {})
+    @registry = args[:registry] || Set.new
```
Comment (paulghaddad) [inline robot_name/registry.rb]: Excellent use case for `Set`!

File: robot_name/robot_name.rb [hunk 16]
```diff
@@ -1,32 +1,93 @@
...
+    new_name = ""
+
+    2.times { new_name.concat(('A'..'Z').to_a.sample) }
+    3.times { new_name.concat(rand(10).to_s) }
+
+    new_name
+  end
+
+  def valid_name?(name)
+    name =~ /[[:alpha:]]{2}[[:digit:]]{3}/
```
Comment (paulghaddad) [inline robot_name/robot_name.rb]: What do you think about making this regex a constant.

**Comment by paulghaddad** [comment]
@DejaAugustine Great job on this exercise. This is one of the best solutions to this exercise I've seen. Take a look at my comments, and let me know your thoughts. 

You are good on this exercise.

File: robot_name/robot_name.rb (hunk 9 above)
Comment (DejaAugustine) [inline robot_name/robot_name.rb]: I agree. Doing so would also do more to document what they do.

**Pull request closed without merging**